advanced multithreading

JohnnyBraver opened the pull request:
found accuracy issues, the main update will wait

JohnnyBraver committed:
basic multithread from reverted branch

found accuracy issues, the main update will wait

yoniinoy commented:
this continues #3

JohnnyBraver committed:
fix clusted detection error

the initial point was being skipped during the search, causing small outlines to be discarded or overlap with the first point

JonathanBraverDev commented:
I've accepted the fact that manual, pre c++11 is unnecessarily complex and will proceed with a multhtread update in the reasonably near future.

JohnnyBraver committed:
tiny refactor

handling each cell update in a thread has HORRIBLE, ABSOLUTLY HORRIBLE so i wont even post that failed commit i accidently puched to main (you'll find it if you really want to)
this update allows to save the seed used for the initial state for better testing in the future, and this version of multithreading ACTUALLY runs faster than the single threaded version.
also fixed indentation in the main

JohnnyBraver committed:
WIP - atomic status resturcute

began to change the way all the aproches update the matrix, focosuing on cutting multiple updates to 0.
single threaded paroach will make better use of its 'old_backup' matrix while the multithreaded one will use a similar atomic mathed that will hopefully not ruin preformance.

JohnnyBraver committed:
Update main.cpp

JohnnyBraver committed:
wip - optimization and fixes

RIP last commit message, this one wil cover both as best as it can.
took too long for what i changed, the way 'old_copy' was used was rather odd.
updated matrix functions and changed the way cells are updated.
also light naming of input variables and use of defined words for clarity.

JohnnyBraver committed:
wip - optimized matrix operations

added a proper copy function, remaned original to clone and used accordingly, trying to minimize redundant deletion/recreation combos

JohnnyBraver committed:
wip - single thread 'FindCluster' update

'FindCluster' now uses the global update map to a void update doubling, accuracy looks fine.
possibley fixing an undiscoved bug in the function. (line 493 -> 501)
also capitalized the global map names
